### PR TITLE
Style checking for core

### PR DESCRIPTION
**Pull request by chidozieononiwu**
Move .editorconfig to Repo Root.
Run dotnet format for Azure.Core to fix style errors.
Turn on Style Cop for Azure.Core
Fix Style Cop errors for Azrue.Core

### CODE REVIEW

File: sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs [hunk 1]
```diff
@@ -14,9 +14,11 @@ internal static class AzureBaseBuffersExtensions
     {
         public static async Task WriteAsync(this Stream stream, ReadOnlyMemory<byte> buffer, CancellationToken cancellation = default)
         {
-            if (stream == null) throw new ArgumentNullException(nameof(stream));
+            if (stream == null)
+                throw new ArgumentNullException(nameof(stream));
 
-            if (buffer.Length == 0) return;
```
Comment (pakrym) [inline sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs]: what rule controls this. I feel this should be fine.

File: sdk/core/Azure.Core/src/Pipeline/HttpPipelineRequestContent.cs [hunk 2]
```diff
@@ -56,7 +57,8 @@ public override void WriteTo(Stream stream, CancellationToken cancellationToken)
                     {
                         cancellationToken.ThrowIfCancellationRequested();
                         var read = _stream.Read(buffer, 0, buffer.Length);
-                        if (read == 0) { break; }
+                        if (read == 0)
+                        { break; }
```
Comment (pakrym) [inline sdk/core/Azure.Core/src/Pipeline/HttpPipelineRequestContent.cs]: This change is not great.

**Review by pakrym — CHANGES_REQUESTED** [review]
Looks good except the newline for single line blocks.

File: sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs (hunk 1 above)
Comment (chidozieononiwu) [reply sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs]: That should be 
csharp_preserve_single_line_statements = false # vs-default: true
Should I, I'll go ahead and set that tot true.

Comment (pakrym) [reply sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs]: Isn't it csharp_preserve_single_line_blocks ?

Comment (chidozieononiwu) [reply sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs]: Well they are similar, but this particular case is  cause by `csharp_preserve_single_line_statements = false`. Actually `csharp_preserve_single_line_blocks = true` is set to true.

Comment (pakrym) [reply sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs]: You are right.

**Comment by pakrym** [comment]
Also, do we want `.editorconfig` at the root? it would affect all projects both track1 and management

**Comment by chidozieononiwu** [comment]
> Also, do we want .editorconfig at the root? it would affect all projects both track1 and management

I discussed this with @heaths and he thought it won't matter since the rules are only applied when one runs dotnet format or used the editorconfig plugin for vscode. For the various other editors that have native support for editorconfig I guess it will still be an issue. The other options will be to keep .editorconfig inside the service folder.

**Comment by pakrym** [comment]
Just doing format document in VS would apply our styles to their code.

**Comment by chidozieononiwu** [comment]
> Just doing format document in VS would apply our styles to their code.

Maybe we should keep .editorconfig inside the package folders since there seems to be no way to add exclusion or scope in the root .editorconfig file

**Comment by pakrym** [comment]
> Maybe we should keep .editorconfig inside the package folders since there seems to be no way to add exclusion or scope in the root .editorconfig file

I think it would be the safest option for now, we might need a script to syncronize them later.

File: sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs (hunk 1 above)
Comment (pakrym) [reply sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs]: Let's keep it as is and just fix the worst offenders like:
``` C#
if (read == 0)
{ break; }
```

**Comment by heaths** [comment]
Another option might be to try path wildcard matching for styles, like:
```
[sd/*/Azure.*/**]
# General files

[sdk/*/Azure.*/**.cs]
# C# styles
```

File: sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs (hunk 1 above)
Comment (pakrym) [reply sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs]: We should re-sync with corefx an pull in new things like `csharp_prefer_braces`.

**Comment by chidozieononiwu** [comment]
> Another option might be to try path wildcard matching for styles, like:
> [sd/*/Azure.*/**]
> # General files
> 
> [sdk/*/Azure.*/**.cs]
> # C# styles

I did try that. It didn't make a difference

> Another option might be to try path wildcard matching for styles, like:
> [sd/*/Azure.*/**]
> # General files
> 
> [sdk/*/Azure.*/**.cs]
> # C# styles

I wish that did the trick, but it still doesn't work.

**Comment by heaths** [comment]
What exactly didn't work? It still applied styles to track 1 sources?

**Comment by chidozieononiwu** [comment]
> What exactly didn't work? It still applied styles to track 1 sources?

Yes, it still runs style check on Track 1, and Management. 
It seems that overides can be applied using the method you suggested. But that will require overriding too many paths and you have to do it one at a time which will mean a bulky .editorconfig file.

**Comment by heaths** [comment]
I just tried an experiment where I put the following content in .editorconfig in the repo root and it worked, i.e. *.cs files under sdk/keyvault/Azure.* used tabs (ick), but when I even just opened an individual file from sdk/keyvault/Microsoft.* it still used spaces.
```
root = true

[sdk/*/Azure.*/**]
indent_style = space
indent_size = 4
trim_trailing_whitespace = true
insert_final_newline = true

[sdk/*/Azure.*/**/*.cs]
dotnet_sort_system_directives_first = true
dotnet_separate_import_directive_groups = false
```

**Comment by chidozieononiwu** [comment]
> I just tried an experiment where I put the following content in .editorconfig in the repo root and it worked, i.e. *.cs files under sdk/keyvault/Azure.* used tabs (ick), but when I even just opened an individual file from sdk/keyvault/Microsoft.* it still used spaces.
> root = true
> 
> [sdk/*/Azure.*/**]
> indent_style = space
> indent_size = 4
> trim_trailing_whitespace = true
> insert_final_newline = true
> 
> [sdk/*/Azure.*/**/*.cs]
> dotnet_sort_system_directives_first = true
> dotnet_separate_import_directive_groups = false

I'll try that again.

**Comment by heaths** [comment]
Also, what editor are you using? I used Visual Studio 2019 that supports .editorconfig natively. I'd be surprised, though, if some editor/extension supported enough of the pattern to apply to certain files but not within a directory. That would be a bug, given https://editorconfig.org documents the pattern as supported.

**Comment by chidozieononiwu** [comment]
> Also, what editor are you using? I used Visual Studio 2019 that supports .editorconfig natively. I'd be surprised, though, if some editor/extension supported enough of the pattern to apply to certain files but not within a directory. That would be a bug, given https://editorconfig.org documents the pattern as supported.

Actually I was testing by running dotnet format which applied the editorconfig rules regardless of the specified path.

**Comment by heaths** [comment]
In my experience, it's pretty buggy. But what's the desired result here? If we just want to get track 2 libraries on par with .editorconfig we could run `dotnet format` on just those initially. Any linting we should do at save/build time anyway, so perhaps something could be put into those project files (or something they import, targeting those client projects). Put another way, I don't see much point of running `dotnet format` more than once initially, so you could do it on those 4 services in track 2.

```powershell
dir sdk/*/Azure.* -recurse -filter *.csproj | foreach { dotnet format -w $_.fullname }
```

After that, conforming editors would at least suggest - if not err, depending on rule severity - changes when editing files. That's all .editorconfig is good for, really. It's violations won't break the build. For that, you need Roslyn Analyzers / StyleCop (the new ones, which are RAs).

**Comment by chidozieononiwu** [comment]
> In my experience, it's pretty buggy. But what's the desired result here? If we just want to get track 2 libraries on par with .editorconfig we could run dotnet format on just those initially. Any linting we should do at save/build time anyway, so perhaps something could be put into those project files (or something they import, targeting those client projects). Put another way, I don't see much point of running dotnet format more than once initially, so you could do it on those 4 services in track 2.
> dir sdk/*/Azure.* -recurse -filter *.csproj | foreach { dotnet format -w $_.fullname }
> After that, conforming editors would at least suggest - if not err, depending on rule severity - changes when editing files. That's all .editorconfig is good for, really. It's violations won't break the build. For that, you need Roslyn Analyzers / StyleCop (the new ones, which are RAs).

You are right. dotnet format is only for the initial code fixes after which I shouldn't be used. I will test again and then likely to use the paths as you suggested.

File: sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs (hunk 1 above)
Comment (chidozieononiwu) [reply sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs]: Do you want me to add all the rules that don't currently exist in our own .editorconfig?

Comment (pakrym) [reply sdk/core/Azure.Core/src/Buffers/StreamMemoryExtensions.cs]: Not as part of this PR, but later - yes.

File: sdk/core/Azure.Core/src/Shared/NullableAttributes.cs [hunk 3]
```diff
@@ -105,7 +109,8 @@ sealed class NotNullIfNotNullAttribute : Attribute
 #else
     public
 #endif
-        sealed class DoesNotReturnAttribute : Attribute { }
+        sealed class DoesNotReturnAttribute : Attribute
```
Comment (pakrym) [inline sdk/core/Azure.Core/src/Shared/NullableAttributes.cs]: @chidozieononiwu can you please also revert these occurrences of unfortunate formatting?

Comment (chidozieononiwu) [reply sdk/core/Azure.Core/src/Shared/NullableAttributes.cs]: Sorry I missed those. Its's reverted now.

**Review by pakrym — APPROVED** [review]

**Pull request merged**